File: ios/precise_compass/Sources/precise_compass/PreciseCompassPlugin.swift
// MIT License — part of precise_compass.
import CoreLocation
import CoreMotion
import Flutter
import UIKit

/// Acquires heading from CoreLocation (and pitch/roll from CoreMotion) and emits
/// a versioned payload over an event channel. All fusion/accuracy/calibration
/// logic lives in pure Dart; this layer only delivers raw native signals,
/// including the continuous `CLHeading.headingAccuracy` and the magnetic-field
/// magnitude derived from `CLHeading.{x,y,z}`.
public class PreciseCompassPlugin: NSObject, FlutterPlugin, FlutterStreamHandler,
  CLLocationManagerDelegate
{
  private var eventSink: FlutterEventSink?
  private let locationManager = CLLocationManager()
  private let motionManager = CMMotionManager()
  private var suppressCalibrationHud = false

  override init() {
    super.init()
    locationManager.delegate = self
    locationManager.headingFilter = 1
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = PreciseCompassPlugin()
    let events = FlutterEventChannel(
      name: "precise_compass/events", binaryMessenger: registrar.messenger())
    events.setStreamHandler(instance)
    let methods = FlutterMethodChannel(
      name: "precise_compass/methods", binaryMessenger: registrar.messenger())
    registrar.addMethodCallDelegate(instance, channel: methods)
  }

  // MARK: - MethodChannel

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getCapabilities":
      result(capabilities())
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func capabilities() -> [String: Any] {
    return [
      "hasRotationVector": motionManager.isDeviceMotionAvailable,
      "hasGeomagneticRotationVector": false,
      "hasMagnetometer": motionManager.isMagnetometerAvailable
        || CLLocationManager.headingAvailable(),
      "hasGyroscope": motionManager.isGyroAvailable,
      "hasAccelerometer": motionManager.isAccelerometerAvailable,
      "supportsTrueHeading": CLLocationManager.headingAvailable(),
    ]
  }

  // MARK: - EventChannel

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink)
    -> FlutterError?
  {
    eventSink = events
    if let args = arguments as? [String: Any] {
      suppressCalibrationHud = args["suppressIosCalibrationHud"] as? Bool ?? false
      if let rate = args["rate"] as? String {
        locationManager.headingFilter = headingFilter(for: rate)
      }
    }
    guard CLLocationManager.headingAvailable() else {
      events(unavailablePayload())
      return nil
    }
    if motionManager.isDeviceMotionAvailable {
      motionManager.deviceMotionUpdateInterval = 1.0 / 30.0
      motionManager.startDeviceMotionUpdates()
    }
    locationManager.startUpdatingHeading()
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    locationManager.stopUpdatingHeading()
    motionManager.stopDeviceMotionUpdates()
    eventSink = nil
    return nil
  }

  // MARK: - CLLocationManagerDelegate

  public func locationManager(
    _ manager: CLLocationManager, didUpdateHeading newHeading: CLHeading
  ) {
    guard let sink = eventSink else { return }
    let offset = interfaceOrientationOffset()

    var payload: [String: Any] = [
      "source": 3,  // platformHeading
      "magneticHeading": normalize(newHeading.magneticHeading + offset),
      "accuracyDegrees": newHeading.headingAccuracy,  // < 0 => Dart treats unknown
      "timestamp": Int(Date().timeIntervalSince1970 * 1000),
    ]
    if newHeading.trueHeading >= 0 {
      payload["trueHeading"] = normalize(newHeading.trueHeading + offset)
    }
    let magnitude = (newHeading.x * newHeading.x + newHeading.y * newHeading.y
      + newHeading.z * newHeading.z).squareRoot()
    if magnitude > 0 {
      payload["magneticFieldMagnitude"] = magnitude
    }
    if let attitude = motionManager.deviceMotion?.attitude {
      payload["pitch"] = attitude.pitch * 180.0 / Double.pi
      payload["roll"] = attitude.roll * 180.0 / Double.pi
    }
    sink(payload)
  }

  /// Driven by `CompassConfig.suppressIosCalibrationHud`: returning `true` lets
  /// iOS present its polished figure-8 calibration HUD.
  public func locationManagerShouldDisplayHeadingCalibration(_ manager: CLLocationManager) -> Bool {
    return !suppressCalibrationHud
  }

  // MARK: - Helpers

  private func headingFilter(for rate: String) -> CLLocationDegrees {
    switch rate {
    case "fastest": return kCLHeadingFilterNone
    case "ui": return 1
    case "normal": return 2
    case "batterySaving": return 5
    default: return 1
    }
  }

  private func interfaceOrientationOffset() -> Double {
    var orientation: UIInterfaceOrientation = .portrait
    if #available(iOS 13.0, *) {
      orientation =
        UIApplication.shared.connectedScenes
        .compactMap { $0 as? UIWindowScene }
        .first?.interfaceOrientation ?? .portrait
    } else {
      orientation = UIApplication.shared.statusBarOrientation
    }
    switch orientation {
    case .portraitUpsideDown: return 180
    case .landscapeLeft: return -90
    case .landscapeRight: return 90
    default: return 0
    }
  }

  private func normalize(_ degrees: Double) -> Double {
    let remainder = degrees.truncatingRemainder(dividingBy: 360)
    return remainder < 0 ? remainder + 360 : remainder
  }

  private func unavailablePayload() -> [String: Any] {
    return [
      "source": -1,
      "accuracyDegrees": -1.0,
      "timestamp": Int(Date().timeIntervalSince1970 * 1000),
    ]
  }
}

File: android/src/main/kotlin/com/precisecompass/precise_compass/PreciseCompassPlugin.kt
/*
 * MIT License — part of precise_compass.
 */
package com.precisecompass.precise_compass

import android.content.Context
import android.hardware.GeomagneticField
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.hardware.display.DisplayManager
import android.location.Location
import android.location.LocationManager
import android.view.Display
import android.view.Surface
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Acquires raw heading data and emits a versioned payload over an event
 * channel. All sensor fusion / accuracy / calibration logic lives in pure Dart;
 * this layer only delivers the highest-fidelity raw signals the device offers,
 * including [SensorEvent.values]`[4]` (estimated heading accuracy, radians) and
 * the magnetic-field magnitude used for interference detection.
 */
class PreciseCompassPlugin :
    FlutterPlugin,
    EventChannel.StreamHandler,
    MethodChannel.MethodCallHandler {

    private var eventChannel: EventChannel? = null
    private var methodChannel: MethodChannel? = null
    private var applicationContext: Context? = null
    private var sensorManager: SensorManager? = null
    private var display: Display? = null

    private var rotationSensor: Sensor? = null
    private var geomagneticRotationSensor: Sensor? = null
    private var accelerometer: Sensor? = null
    private var magnetometer: Sensor? = null
    private var gyroscope: Sensor? = null

    private var listener: SensorEventListener? = null
    private var eventSink: EventChannel.EventSink? = null

    private var activeVectorSensor: Sensor? = null
    private var activeSourceCode: Int = SOURCE_UNAVAILABLE
    private var samplingPeriodUs: Int = DEFAULT_PERIOD_US
    private var declinationDegrees: Double? = null

    private val gravity = FloatArray(3)
    private val geomagnetic = FloatArray(3)
    private var hasGravity = false
    private var hasGeomagnetic = false
    private var fieldMagnitude: Double? = null
    private var osAccuracyStatus: Int = -1

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        sensorManager =
            applicationContext?.getSystemService(Context.SENSOR_SERVICE)
                as? SensorManager
        display = (applicationContext?.getSystemService(Context.DISPLAY_SERVICE)
            as? DisplayManager)?.getDisplay(Display.DEFAULT_DISPLAY)

        sensorManager?.let { sm ->
            rotationSensor = sm.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)
            geomagneticRotationSensor =
                sm.getDefaultSensor(Sensor.TYPE_GEOMAGNETIC_ROTATION_VECTOR)
            accelerometer = sm.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
            magnetometer = sm.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD)
            gyroscope = sm.getDefaultSensor(Sensor.TYPE_GYROSCOPE)
        }

        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL).apply {
            setStreamHandler(this@PreciseCompassPlugin)
        }
        methodChannel =
            MethodChannel(binding.binaryMessenger, METHOD_CHANNEL).apply {
                setMethodCallHandler(this@PreciseCompassPlugin)
            }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        unregister()
        eventChannel?.setStreamHandler(null)
        methodChannel?.setMethodCallHandler(null)
        eventChannel = null
        methodChannel = null
        sensorManager = null
        applicationContext = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getCapabilities" -> result.success(capabilities())
            else -> result.notImplemented()
        }
    }

    private fun capabilities(): Map<String, Any> {
        val hasRv = rotationSensor != null
        val hasGeo = geomagneticRotationSensor != null
        val hasMag = magnetometer != null
        val hasAccel = accelerometer != null
        return mapOf(
            "hasRotationVector" to hasRv,
            "hasGeomagneticRotationVector" to hasGeo,
            "hasMagnetometer" to hasMag,
            "hasGyroscope" to (gyroscope != null),
            "hasAccelerometer" to hasAccel,
            "supportsTrueHeading" to (hasRv || hasGeo || (hasMag && hasAccel)),
        )
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        configure(arguments)
        if (!register()) {
            events?.success(unavailablePayload())
        }
    }

    override fun onCancel(arguments: Any?) {
        unregister()
        eventSink = null
    }

    @Suppress("UNCHECKED_CAST")
    private fun configure(arguments: Any?) {
        val args = arguments as? Map<String, Any?> ?: emptyMap()
        samplingPeriodUs = when (args["rate"] as? String) {
            "fastest" -> 5_000
            "ui" -> 16_000
            "normal" -> 33_000
            "batterySaving" -> 100_000
            else -> DEFAULT_PERIOD_US
        }
        selectSource(args["fusionMode"] as? String ?: "auto")
        declinationDegrees = computeDeclination()
    }

    private fun selectSource(mode: String) {
        activeVectorSensor = when (mode) {
            "geomagnetic" -> geomagneticRotationSensor ?: rotationSensor
            "rotationVector" -> rotationSensor ?: geomagneticRotationSensor
            else -> rotationSensor ?: geomagneticRotationSensor
        }
        activeSourceCode = when {
            activeVectorSensor?.type == Sensor.TYPE_ROTATION_VECTOR ->
                SOURCE_ROTATION_VECTOR
            activeVectorSensor != null -> SOURCE_GEOMAGNETIC
            accelerometer != null && magnetometer != null -> SOURCE_GEOMAGNETIC
            else -> SOURCE_UNAVAILABLE
        }
    }

    private fun register(): Boolean {
        val sm = sensorManager ?: return false
        val newListener = CompassListener()
        listener = newListener
        val vector = activeVectorSensor
        return when {
            vector != null -> {
                sm.registerListener(newListener, vector, samplingPeriodUs)
                magnetometer?.let {
                    sm.registerListener(newListener, it, samplingPeriodUs)
                }
                true
            }
            accelerometer != null && magnetometer != null -> {
                sm.registerListener(newListener, accelerometer, samplingPeriodUs)
                sm.registerListener(newListener, magnetometer, samplingPeriodUs)
                true
            }
            else -> false
        }
    }

    private fun unregister() {
        listener?.let { sensorManager?.unregisterListener(it) }
        listener = null
        hasGravity = false
        hasGeomagnetic = false
        fieldMagnitude = null
        osAccuracyStatus = -1
    }

    private fun computeDeclination(): Double? {
        val ctx = applicationContext ?: return null
        return try {
            val lm = ctx.getSystemService(Context.LOCATION_SERVICE)
                as? LocationManager ?: return null
            val providers = listOf(
                LocationManager.GPS_PROVIDER,
                LocationManager.NETWORK_PROVIDER,
                LocationManager.PASSIVE_PROVIDER,
            )
            var best: Location? = null
            for (provider in providers) {
                val location = lm.getLastKnownLocation(provider) ?: continue
                if (best == null || location.time > best.time) best = location
            }
            best?.let {
                GeomagneticField(
                    it.latitude.toFloat(),
                    it.longitude.toFloat(),
                    it.altitude.toFloat(),
                    it.time,
                ).declination.toDouble()
            }
        } catch (e: SecurityException) {
            null // No location permission: magnetic heading only.
        } catch (e: Exception) {
            null
        }
    }

    private fun displayRotation(): Int = display?.rotation ?: Surface.ROTATION_0

    private inner class CompassListener : SensorEventListener {
        override fun onSensorChanged(event: SensorEvent) {
            when (event.sensor.type) {
                Sensor.TYPE_ROTATION_VECTOR,
                Sensor.TYPE_GEOMAGNETIC_ROTATION_VECTOR -> {
                    if (event.sensor == activeVectorSensor) emitFromVector(event)
                }
                Sensor.TYPE_MAGNETIC_FIELD -> {
                    fieldMagnitude = CompassMath.magnitude(event.values).toDouble()
                    if (activeVectorSensor == null) {
                        System.arraycopy(event.values, 0, geomagnetic, 0, 3)
                        hasGeomagnetic = true
                        emitFromAccelMag()
                    }
                }
                Sensor.TYPE_ACCELEROMETER -> {
                    if (activeVectorSensor == null) {
                        System.arraycopy(event.values, 0, gravity, 0, 3)
                        hasGravity = true
                        emitFromAccelMag()
                    }
                }
            }
        }

        override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {
            if (sensor.type == Sensor.TYPE_MAGNETIC_FIELD) {
                osAccuracyStatus = accuracy
            }
        }
    }

    private fun emitFromVector(event: SensorEvent) {
        val orientation = CompassMath.orientationFromRotationVector(
            CompassMath.safeRotationVector(event.values),
            displayRotation(),
        )
        // values[4] = estimated heading accuracy in radians (rotation vector).
        val accuracyDegrees =
            if (event.values.size >= 5 && event.values[4] >= 0f) {
                Math.toDegrees(event.values[4].toDouble())
            } else {
                -1.0
            }
        emit(CompassMath.azimuthDegrees(orientation.azimuthRadians), accuracyDegrees, orientation)
    }

    private fun emitFromAccelMag() {
        if (!hasGravity || !hasGeomagnetic) return
        val orientation = CompassMath.orientationFromAccelMag(
            gravity,
            geomagnetic,
            displayRotation(),
        ) ?: return
        emit(CompassMath.azimuthDegrees(orientation.azimuthRadians), -1.0, orientation)
    }

    private fun emit(
        magneticHeading: Double,
        accuracyDegrees: Double,
        orientation: Orientation,
    ) {
        val sink = eventSink ?: return
        if (magneticHeading.isNaN()) return
        val payload = HashMap<String, Any?>()
        payload["source"] = activeSourceCode
        payload["magneticHeading"] = magneticHeading
        declinationDegrees?.let {
            payload["trueHeading"] = (magneticHeading + it + 360.0) % 360.0
        }
        payload["accuracyDegrees"] = accuracyDegrees
        payload["osAccuracyStatus"] = osAccuracyStatus
        fieldMagnitude?.let { payload["magneticFieldMagnitude"] = it }
        payload["pitch"] = Math.toDegrees(orientation.pitchRadians.toDouble())
        payload["roll"] = Math.toDegrees(orientation.rollRadians.toDouble())
        payload["timestamp"] = System.currentTimeMillis()
        sink.success(payload)
    }

    private fun unavailablePayload(): Map<String, Any?> = mapOf(
        "source" to SOURCE_UNAVAILABLE,
        "accuracyDegrees" to -1.0,
        "timestamp" to System.currentTimeMillis(),
    )

    companion object {
        private const val EVENT_CHANNEL = "precise_compass/events"
        private const val METHOD_CHANNEL = "precise_compass/methods"
        private const val DEFAULT_PERIOD_US = 16_000
        private const val SOURCE_ROTATION_VECTOR = 0
        private const val SOURCE_GEOMAGNETIC = 1
        private const val SOURCE_UNAVAILABLE = -1
    }
}
